title_bar: Add configurable window controls position

sousaakira opened the pull request:
Release Notes:

- Added the ability to customize in which side (left or right) the windows controls are rendered on Linux.

---

## 🎯 Description
Adds configurable window control buttons (minimize, maximize, close) positioning for Linux, allowing users to choose between macOS-style (left side) or Windows-style (right side) placement.

## ✨ Features
- New `title_bar.window_controls_position` setting with `"left"` and `"right"` options
- Left positioning: macOS style (Close → Minimize → Maximize)
- Right positioning: Windows style (Minimize → Maximize → Close)
- Fixed transparent background issues for window controls
- Maintains consistent styling with Zed's theme

## 🔧 Technical Changes

### Settings System
- Added `WindowControlsPosition` enum in `settings_content.rs`
- Extended `TitleBarSettingsContent` with `window_controls_position` field
- Updated `TitleBarSettings` to include the new configuration

### Title Bar Layout
- Modified `platform_title_bar.rs` to use setting for layout positioning
- Added conditional logic for `justify_start()` vs `justify_between()` based on position
- Fixed transparent container background by adding `bg(titlebar_color)`

### Window Controls
- Updated `platform_linux.rs` to reorder buttons based on position setting
- Changed button background from `ghost_element_background` to `title_bar_background`
- Implemented proper button sequencing for both positions

## 🧪 How to Test
1. Add to your Zed settings:
   ```json
   {
     "title_bar": {
       "window_controls_position": "left"
     }
   }
   ```
   or
   ```json
   {
     "title_bar": {
       "window_controls_position": "right"
     }
   }
   ```
2. Restart Zed
3. Verify buttons are positioned correctly
4. Check that background is not transparent
5. Test button functionality (minimize, maximize, close)

## �� Expected Behavior
- **Left position**: Buttons appear on the left side of the title bar in Close → Minimize → Maximize order
- **Right position**: Buttons appear on the right side of the title bar in Minimize → Maximize → Close order
- **Background**: Solid background matching Zed's theme (no transparency)

## 🔍 Files Changed
- `crates/settings/src/settings_content.rs` - Added enum and setting
- `crates/title_bar/src/title_bar_settings.rs` - Updated settings struct
- `crates/title_bar/src/platform_title_bar.rs` - Modified layout logic
- `crates/title_bar/src/platforms/platform_linux.rs` - Updated button ordering and styling

## 🎨 Design Rationale
This feature provides Linux users with the flexibility to choose their preferred window control button layout, improving the user experience by allowing them to match their desktop environment's conventions or personal preferences.

## ✅ Checklist
- [x] Code compiles without errors
- [x] Settings are properly serialized/deserialized
- [x] Background transparency issues resolved
- [x] Button ordering works correctly for both positions
- [x] Layout adapts properly based on configuration
- [x] No breaking changes to existing functionality

## 🔗 Related
This addresses the need for customizable window control positioning on Linux, providing consistency with user expectations from different desktop environments.

![demo2](https://github.com/user-attachments/assets/7333db34-d54e-427c-ac52-140925363f91)

sousaakira committed:
title_bar: Ajustar comportamento de justificação no Linux e corrigir controle de janela

Modificações realizadas:
- Ajustada a justificação do layout no `PlatformTitleBar` para que utilize `justify_start` no estilo Linux, em vez de `justify_between`.
- Alterado o controle de janela no `LinuxWindowControls` para usar `WindowControl::new_close` para o botão de fechar e `WindowControl::new` para o botão de minimizar.

Release Notes:
- N/A

sousaakira committed:
Merge branch 'main' of github.com:sousaakira/zed

sousaakira committed:
feat: add configurable window controls position for Linux

- Add WindowControlsPosition enum with Left/Right options
- Add window_controls_position setting to TitleBarSettingsContent
- Update TitleBarSettings to include new configuration
- Modify platform_title_bar.rs to use setting for layout positioning
- Update platform_linux.rs to reorder buttons based on position:
  * Left: Close -> Minimize -> Maximize (macOS style)
  * Right: Minimize -> Maximize -> Close (Windows style)
- Fix transparent background issues by using title_bar_background
- Ensure proper background for left-positioned controls container

Closes #[ISSUE_NUMBER]

sousaakira commented:
@cla-bot check

danilo-leal commented:
Hey, thanks for the PR! Can we get some screenshots of how this affects the UI? I don't have a Linux machine easy with me so I can experiment with it.

sousaakira committed:
refactor: improve window controls position implementation

- Change default position from 'left' to 'right' for better Windows compatibility
- Add comprehensive docstring explaining valid values ('left' and 'right')
- Implement Default trait for WindowControlsPosition enum
- Refactor LinuxWindowControls to reduce code duplication with build_controls helper
- Remove unnecessary unit tests to keep codebase clean

These improvements make the code more maintainable and user-friendly.

danilo-leal reviewed (approved):
Thanks for the screenshots! Looks good to me.

danilo-leal commented:
@sousaakira There's a CI error there with the tests import; can you verify?

Be-ing commented on crates/title_bar/src/platform_title_bar.rs:
```diff
@@ -132,35 +144,76 @@ impl Render for PlatformTitleBar {
+                            let title_bar_settings = TitleBarSettings::get(None, cx);
+                            match title_bar_settings.window_controls_position {
+                                WindowControlsPosition::Left => {
+                                    // macOS style: controles no início da title bar
```
Portuguese comment?

Be-ing commented on crates/title_bar/src/platform_title_bar.rs:
```diff
@@ -132,35 +144,76 @@ impl Render for PlatformTitleBar {
+                                        )
+                                }
+                                WindowControlsPosition::Right => {
+                                    // Windows style: controles no final da title bar
```
Portuguese comment?

Be-ing commented on crates/settings/src/settings_content.rs:
```diff
@@ -234,6 +234,21 @@ pub enum BaseKeymapContent {
     None,
 }
 
+/// Position of window control buttons on Linux.
+///
+/// Valid values: "left" or "right"
+/// - "left": Window controls on the left side (macOS style)
+/// - "right": Window controls on the right side (Windows style)
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, JsonSchema, MergeFrom, PartialEq, Eq, Default)]
+#[serde(rename_all = "snake_case")]
+pub enum WindowControlsPosition {
+    /// Window controls on the left side (macOS style)
+    Left,
+    /// Window controls on the right side (Windows style)
+    #[default]
+    Right,
+}
+
```
Instead of making the user manually configure this in Zed's settings.json, this could be handled automatically by querying GSettings with the [gio crate](https://gtk-rs.org/gtk-rs-core/stable/latest/docs/gio/struct.Settings.html). The setting only impacts GNOME (even with GTK + libadwaita applications), so when Zed is not running on GNOME, the window controls should always be drawn on the right.

I can make a pull request for your branch implementing this.

danilo-leal committed:
Merge remote-tracking branch 'origin/main' into pr/38834

danilo-leal committed:
Add code back after merge with main

danilo-leal committed:
Fix clippy

Be-ing commented:
I recommend reverting this ASAP before users start using it. Adding settings to Zed for this is not a good way to do it. Zed should read the system's settings, not add its own that would have to be manually configured to match the system. I started implementing that in #39395, though my focus is currently on https://github.com/zed-industries/zed/issues/28754.
